plugin/core/windows.py

```diff
@@ -557,6 +566,48 @@ def _check_window_closed(self) -> None:
         if not self._is_closing and not self._window.is_valid():
             self._handle_window_closed()
 
+    def _receive_progress_token(self, params: Dict[str, Any], client: Client, request_id: Any) -> None:
+        token = params.get('token')  # number | string
+        if token:
+            self._progress[token] = dict()
+        client.send_response(Response(request_id, None))
+
+    def _show_progress(self, params: Dict[str, Any]) -> None:
+        token = params.get('token')
+        value = params.get('value')
+        if token and value:
+            if value.get('kind') == 'begin':
+                if token not in self._progress:
+                    self._progress[token] = dict()
+                self._progress[token]['title'] = value.get('title')  # mandatory
```

Comment:
`$/progress` notification is very generic and nothing is mandatory in the `value`.

You are assuming here that progress notifications are all workDoneProgress related.

The proper solution is to only handle progress notifications for tokens that were created through `window/workDoneProgress/create` and ignore any other (as we don't know how to handle them - that would be something for server-specific packages to implement).

Also, since most (if not all) requests inherit from `WorkDoneProgressParams`, that means all of them can provide token for that particular request. That is also not handled (but might not need to be if it's not needed for Julia right now).

Reply:
The type of the `value` field depends on what the original request was. In the case of `window/workDoneProgress/create` it is server-initiated. This is, from what I understand, the exception to the rule.

What would generally happen is that the editor does a request to the server, say `textDocument/references`, and passes along a `workDoneToken`.

Then, the server sends a bunch of `$/progress` notifications using the `workDoneToken` supplied by editor. The type of `value` is then always what is assumed here in this diff.

But editor can *also* send a `partialResultToken`. Then, the server is allowed to send partial results via the `$/progress` notification. The `token` would then correspond to the `partialResultToken`, and the type of `value` is whatever the type is of the original request from editor (in this example it would be partial reference results from `textDocument/references`).

So the `$/progress` notification serves two purposes.

But it doesn't matter for this PR.

Reply:
Yes, so I'm saying the PR should only handle the `$/progress` requests related to the tokens that were either sent with the request (we don't send those right now) or through  `window/workDoneProgress/create`. Ignore all other progress notifications.

And rename `_show_progress` to, for example, `_handle_work_progress` to make the logic more explicit.

(I wonder what Julia is doing and if that would be enough to support it...)

Reply:
Yes, this PR can only handle notifications for `workDoneProgress`. As we don't send a `partialResultToken`, this assumption is valid for now. I've added the lines
```python
if token not in self._progress:
    self._progress[token] = dict()
```
for notifications with `"kind": "begin"` so that the following two lines don't throw an error for unknown tokens and this way we could still show notifications if the server somehow forgets to send a `window/workDoneProgress/create` request first. Actually I thought this was the case with `rust-analyzer`, but after checking again I get the following log:
```
:: <-- rust-analyzer window/workDoneProgress/create(1): {'token': 'rustAnalyzer/startup'}
:: >>> rust-analyzer 1: None
:: <-  rust-analyzer $/progress: {'token': 'rustAnalyzer/startup', 'value': {'message': '0/1 packages', 'percentage': 0.0, 'title': 'rust-analyzer', 'kind': 'begin'}}
:: <-  rust-analyzer $/progress: {'token': 'rustAnalyzer/startup', 'value': {'message': 'rust-analyzer loaded, 1 packages', 'kind': 'end'}}
```
Also it still works with the suggestion to use `pop(token, None)` for `"end"` messages: the spec seems a bit unlear, but if I understand it right, it doesn't forbid to send the `window/workDoneProgress/create` request only once, and then reuse that token for multiple sets of notifications with `begin/report/end`.
Should we be more strict and ignore all tokens not submitted via `window/workDoneProgress/create`?

Julia also sends the request first:
```
:: <-- julials window/workDoneProgress/create(faa26dc2-c47f-402d-8490-2fb5d944f63d): {'token': '61cc1844-6988-4178-8aa5-912dbfed892f'}
:: >>> julials faa26dc2-c47f-402d-8490-2fb5d944f63d: None
:: <-  julials $/progress: {'token': '61cc1844-6988-4178-8aa5-912dbfed892f', 'value': {'message': 'Indexing packages...', 'title': 'Julia Language Server', 'kind': 'begin'}}
```

Reply:
> Yes, this PR can only handle notifications for `workDoneProgress`. As we don't send a `partialResultToken`, this assumption is valid for now.

I'm arguing that technically `$/progress` can be sent from the server for whatever reason and with whatever value, unrelated to either work progress or partial results. Not sure why it would do that and probably it's not likely that it would that as how would client know how to handle it, but it's possible.

Reply:
I agree with @rchl,
The `$/progress` notification is to generic.
```
type ProgressToken = number | string;
interface ProgressParams<T> {
	/**
	 * The progress token provided by the client or server.
	 */
	token: ProgressToken;

	/**
	 * The progress data.
	 */
	value: T;
}
```

We can't assume that the value, will have a `title` or a `message`.
For me the `$/progress` notification should be implemented in `LanguageHandlers` (take [LSP-rust](https://github.com/tomv564/LSP-rust/blob/ecad648dfcebaba6e375338e1848aa3079481e0f/plugin.py#L124-L134) for example, although not the $progress notification, but it is similar).

There are some request/notifications that are to generic to be in the core LSP plugin (code lens is another example, it is also to generic),
and should be in a language handler.

Reply:
We've agreed in another thread to handle only "work progress" related notifications which are specified exactly.

Reply:
Talking about this one:
https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workDoneProgressBegin

Missed that part...

Reply:
Yes, we ignore all notifications with unknown `"token"` values now. Tokens can (for now) only be added by the server via an `window/workDoneProgress/create` request, so that we can assume that all notifications with corresponding tokens follow the implemented conventions for the `"value"` object.